### PR TITLE
[s3-repository] Refresh Web Identity Credentials when Web Identity token changes

### LINKED ISSUE #101828: Elasticsearch does not refresh AWS Web Identity Token file when changed on disk
### Elasticsearch Version

8.x

### Installed Plugins

_No response_

### Java Version

_bundled_

### OS Version

Kubernetes 

### Problem Description

**Observed**
We have ECK users reporting that S3 repository setups that rely on AWS IAM roles for service accounts (IRSA) stop working after a while and only a restart of the affected Pods restores functionality. 

https://github.com/elastic/cloud-on-k8s/issues/7208

The users are following this documentation on the ECK side https://www.elastic.co/guide/en/cloud-on-k8s/master/k8s-snapshots.html#k8s-iam-service-accounts which configures a `CustomWebIdentityTokenCredentialsProvider` based on information from the two environment variables `AWS_WEB_IDENTITY_TOKEN_FILE` (and `AWS_ROLE_ARN`) 

My hunch is that Elasticsearch is not watching the file pointed to by `AWS_WEB_IDENTITY_TOKEN_FILE` and does not [refresh](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/auth/STSAssumeRoleWithWebIdentitySessionCredentialsProvider.html#refresh--) the credentials when the token changes on disk after the configured expiration time. This would be consistent with the users reporting that a restart of the Pods solves the problem. 

**Expected** 
I think it would be reasonable to expect that Elasticsearch reloads the credentials when the token changes on disk without user intervention. 

### Steps to Reproduce

https://www.elastic.co/guide/en/cloud-on-k8s/master/k8s-snapshots.html#k8s-iam-service-accounts

with a short ` expirationSeconds: 300` or something along those lines. 
Then wait for token refresh. (Disclaimer: I have not tried it yet myself just relaying user reports here)

### Logs (if relevant)

_No response_

### PR DESCRIPTION
**Pull request by arteam**
`STSAssumeRoleWithWebIdentitySessionCredentialsProvider` is supposed to automatically refresh credential when the current credentials expire and pick up a new web identity token if it has been rotated, but there have been reports from the field that doesn't work reliably. To make sure that the credentials get refreshed, let's create a file watcher that explicitly refresh the credentials when the web identity token changes

Resolves #101828

### CODE REVIEW

**Review by ywangd — COMMENTED** [review]
Thanks for working on the issue. A few high level questions.

1. Do we know why AWS itself does not refresh the credentials? Is it a bug, a configuration issue, or user issue, e.g. token file is updated before the configured session time or something else?
2. Would it be better to implement this with [file watching](https://github.com/elastic/elasticsearch/blob/a0cf690be8aa3b10d0767d7941259cc1ce0db223/server/src/main/java/org/elasticsearch/watcher/FileWatcher.java#L27) instead of periodical refresh on a timer? I think the issue description (#101828) suggests "refresh on file changes".

**Comment by arteam** [comment]
Hi @ywangd!

> Do we know why AWS itself does not refresh the credentials? Is it a bug, a configuration issue, or user issue, e.g. token file is updated before the configured session time or something else?

The AWS SDK is supposed automatically pick up a new token when the session expires, so the process of rotating the token should be transparent. But, apparently when the token gets rotated, the current session becomes stale, so we have to refresh credentials and force the SDK to create a new session with the new token.

> Would it be better to implement this with [file watching](https://github.com/elastic/elasticsearch/blob/a0cf690be8aa3b10d0767d7941259cc1ce0db223/server/src/main/java/org/elasticsearch/watcher/FileWatcher.java#L27) instead of periodical refresh on a timer

That's a great idea! That was my first approach for solving issue, but I thought we would need to create a separate thread for watching the file, but I didn't realize we can leverage `ResourceWatcherService`. I've updated the PR :+1:

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Service.java [hunk 1]
```diff
@@ -395,6 +400,21 @@ static class CustomWebIdentityTokenCredentialsProvider implements AWSCredentials
...
+                    public void onFileChanged(Path file) {
+                        if (file.equals(webIdentityTokenFileSymlink)) {
+                            credentialsProvider.refresh();
+                        }
+                    }
+                });
+                try {
+                    resourceWatcherService.add(watcher, ResourceWatcherService.Frequency.LOW);
+                } catch (IOException e) {
+                    throw new ElasticsearchException("failed to start watching service_tokens file [{}]", e, webIdentityTokenFileSymlink);
```
Comment (ywangd) [inline modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Service.java:416]:
```suggestion
                    throw new ElasticsearchException("failed to start watching AWS web identity token file [{}]", e, webIdentityTokenFileSymlink);
```

File: modules/repository-s3/src/test/java/org/elasticsearch/repositories/s3/CustomWebIdentityTokenCredentialsProviderTests.java [hunk 2]
```diff
@@ -124,8 +157,64 @@ public void testCreateWebIdentityTokenCredentialsProvider() throws Exception {
...
+
+            String newWebIdentityToken = "88f84342080d4671a511e10ae905b2b0";
+            Files.writeString(environment.configFile().resolve("repository-s3/aws-web-identity-token-file"), newWebIdentityToken);
+
+            var latch = new CountDownLatch(1);
+            webIdentityTokenCheck.setDelegate(s -> {
+                if (s.equals(newWebIdentityToken)) {
+                    latch.countDown();
+                }
+            });
```
Comment (ywangd) [inline modules/repository-s3/src/test/java/org/elasticsearch/repositories/s3/CustomWebIdentityTokenCredentialsProviderTests.java:215]: This needs to be placed before updating the token file. There is a chance that the refresh process gets triggered and completes before this block of code can run.

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Service.java [hunk 3]
```diff
@@ -395,6 +400,21 @@ static class CustomWebIdentityTokenCredentialsProvider implements AWSCredentials
...
                     webIdentityTokenFileSymlink.toString()
                 ).withStsClient(stsClient).build();
+                var watcher = new FileWatcher(webIdentityTokenFileSymlink);
+                watcher.addListener(new FileChangesListener() {
+
+                    @Override
+                    public void onFileChanged(Path file) {
+                        if (file.equals(webIdentityTokenFileSymlink)) {
+                            credentialsProvider.refresh();
+                        }
```
Comment (ywangd) [inline modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Service.java:410]: 1. Can we have a debug log here?
2. A deletion plus re-creation is basically an update as well. Therefore I think we should handle `onFileCreated` as well. We can skip `onFileDeleted` if we do not want to actively destroy the session. Or maybe we could just log a message for it.

**Review by ywangd — APPROVED** [review]
LGTM

A few comments that I'd appreciate addressing, but no need for another round. Thanks!

**Comment by inakiabt** [comment]
Will this work with a chain of symlinks? :)

We're restricting the plugin to accessing only what is inside a specific directory, so we need to create a symlink to the actual token.
AWS EKS is storing the token in `/run/secrets/eks.amazonaws.com/serviceaccount`, but it's being symlinked through `..data` and then `token` as can be seen below:

```shell
$ ls -la /run/secrets/eks.amazonaws.com/serviceaccount/
total 0
drwxrwsrwt 3 root 1001 100 Apr  5 19:00 .
drwxr-xr-x 3 root root  28 Apr  5 19:00 ..
drwxr-sr-x 2 root 1001  60 Apr  5 19:00 ..2024_04_05_19_00_30.914006556
lrwxrwxrwx 1 root 1001  31 Apr  5 19:00 ..data -> ..2024_04_05_19_00_30.914006556
lrwxrwxrwx 1 root 1001  12 Apr  5 19:00 token -> ..data/token
```

We are creating a symlink in the plugin folder to `$AWS_WEB_IDENTITY_TOKEN_FILE` that is  `/run/secrets/eks.amazonaws.com/serviceaccount/token` with
```shell
ln -s $AWS_WEB_IDENTITY_TOKEN_FILE /path/to/repository-s3/aws-web-identity-token-file
```

It works for the first hour until the token expires. We were expecting the `credentialsProvider` to be refreshed after the FileWatcher founds the token is changed, but it is not. We've even enabled logging to see if [this](https://github.com/elastic/elasticsearch/blob/5a9a9b87ac78556fd3dd0953630a8c6e96e9ea1c/modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Service.java#L417) is logged, but it is not either.

We are getting out of ideas.

**Comment by inakiabt** [comment]
nvm, I can confirm it detects the file change. It seems to be another issue.

**Comment by stdmje** [comment]
@arteam could you please confirm it this change has been merged in 8.13? I can't find it in the release notes. Thanks!

**Comment by arteam** [comment]
@stdmje Yes, I can confirm that this change was shipped to 8.13. It looks like the automation bot didn't generate a changelog for this change, so it was missed from the release notes.

**Pull request merged**